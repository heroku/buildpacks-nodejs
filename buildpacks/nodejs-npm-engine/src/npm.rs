use commons::fun_run::CmdError;
use libcnb::Env;
use std::process::Command;

#[derive(Debug)]
pub(crate) enum VersionError {
    Command(CmdError),
    Parse(String, heroku_nodejs_utils::vrs::VersionError),
}

pub(crate) struct Version<'a> {
    pub(crate) env: &'a Env,
}

impl<'a> From<Version<'a>> for Command {
    fn from(value: Version<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("--version");
        cmd.envs(value.env);
        cmd
    }
}
