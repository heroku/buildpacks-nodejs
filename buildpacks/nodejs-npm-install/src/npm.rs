use fun_run::CmdError;
use libcnb::Env;
use std::path::PathBuf;
use std::process::Command;

pub(crate) struct SetCacheConfig<'a> {
    pub(crate) env: &'a Env,
    pub(crate) cache_dir: PathBuf,
}

impl<'a> From<SetCacheConfig<'a>> for Command {
    fn from(value: SetCacheConfig<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.args([
            "config",
            "set",
            "cache",
            &value.cache_dir.to_string_lossy(),
            "--global",
        ]);
        cmd.envs(value.env);
        cmd
    }
}

#[derive(Debug)]
pub(crate) enum VersionError {
    Command(CmdError),
    Parse(String, heroku_nodejs_utils::vrs::VersionError),
}
pub(crate) struct Version<'a> {
    pub(crate) env: &'a Env,
}

impl<'a> From<Version<'a>> for Command {
    fn from(value: Version<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("--version");
        cmd.envs(value.env);
        cmd
    }
}

pub(crate) struct Install<'a> {
    pub(crate) env: &'a Env,
}

impl<'a> From<Install<'a>> for Command {
    fn from(value: Install<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("ci");
        cmd.arg("--production=false");
        cmd.envs(value.env);
        cmd
    }
}

pub(crate) struct RunScript<'a> {
    pub(crate) env: &'a Env,
    pub(crate) script: String,
}

impl<'a> From<RunScript<'a>> for Command {
    fn from(value: RunScript<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.args(["run", &value.script]);
        cmd.envs(value.env);
        cmd
    }
}
