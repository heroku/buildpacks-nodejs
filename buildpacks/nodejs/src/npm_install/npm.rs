use fun_run::CmdError;
use libcnb::Env;
use std::path::PathBuf;
use std::process::Command;

pub(crate) struct SetCacheConfig<'a> {
    pub(crate) env: &'a Env,
    pub(crate) cache_dir: &'a PathBuf,
}

impl SetCacheConfig<'_> {
    pub(crate) fn into_command(self) -> Command {
        self.into()
    }
}

impl<'a> From<SetCacheConfig<'a>> for Command {
    fn from(value: SetCacheConfig<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.args([
            "config",
            "set",
            "cache",
            &value.cache_dir.to_string_lossy(),
            "--global",
        ]);
        cmd.envs(value.env);
        cmd
    }
}

#[derive(Debug)]
pub(crate) enum VersionError {
    Command(CmdError),
    Parse(String, heroku_nodejs_utils::vrs::VersionError),
}

pub(crate) struct Version<'a> {
    pub(crate) env: &'a Env,
}

impl Version<'_> {
    pub(crate) fn into_command(self) -> Command {
        self.into()
    }
}

impl<'a> From<Version<'a>> for Command {
    fn from(value: Version<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("--version");
        cmd.envs(value.env);
        cmd
    }
}

pub(crate) struct Install<'a> {
    pub(crate) env: &'a Env,
}

impl Install<'_> {
    pub(crate) fn into_command(self) -> Command {
        self.into()
    }
}

impl<'a> From<Install<'a>> for Command {
    fn from(value: Install<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("ci");
        cmd.envs(value.env);
        cmd
    }
}

pub(crate) struct RunScript<'a> {
    pub(crate) env: &'a Env,
    pub(crate) script: String,
}

impl RunScript<'_> {
    pub(crate) fn into_command(self) -> Command {
        self.into()
    }
}

impl<'a> From<RunScript<'a>> for Command {
    fn from(value: RunScript<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.args(["run", &value.script]);
        cmd.envs(value.env);
        cmd
    }
}

pub(crate) struct Prune<'a> {
    pub(crate) env: &'a Env,
}

impl Prune<'_> {
    pub(crate) fn into_command(self) -> Command {
        self.into()
    }
}

impl<'a> From<Prune<'a>> for Command {
    fn from(value: Prune<'a>) -> Self {
        let mut cmd = Command::new("npm");
        cmd.arg("prune");
        cmd.env("NODE_ENV", "production");
        cmd.envs(value.env);
        cmd
    }
}
