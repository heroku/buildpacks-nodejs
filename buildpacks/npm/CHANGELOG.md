# Changelog
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- `npm ci` and `npm install` now run with `--production=false` to ensure `devDependencies` are installed ([186](https://github.com/heroku/buildpacks-nodejs/pull/186))

## [0.4.4] 2021/06/15
### Fixed
- Clear cache when node version changes ([#40](https://github.com/heroku/buildpacks-node/pull/40))

## [0.4.3] 2021/03/04
- Flush cache when stack image changes ([#28](https://github.com/heroku/buildpacks-node/pull/28))
- Trim whitespace when getting stack name ([#29](https://github.com/heroku/buildpacks-node/pull/29))
- Fail if two lock files are detected ([#30](https://github.com/heroku/buildpacks-node/pull/30))

## [0.4.2] 2021/02/23
- Add license to buildpack.toml ([#17](https://github.com/heroku/buildpacks-node/pull/17))

## [0.4.1] 2021/01/20
- Ensure prefix directory exists ([#42](https://github.com/heroku/nodejs-npm-buildpack/pull/44))
- Use new logging style ([#45](https://github.com/heroku/nodejs-npm-buildpack/pull/45))
- Change log colors to use ANSI codes ([#47](https://github.com/heroku/nodejs-npm-buildpack/pull/47))

## [0.4.0] 2020/11/11
### Added
- Add heroku-20 to supported stacks ([#40](https://github.com/heroku/nodejs-npm-buildpack/pull/40))

## [0.3.0] 2020/09/16
### Added
- Prune devdependencies ([#32](https://github.com/heroku/nodejs-npm-buildpack/pull/32))
- Opt out of pruning devdependencies if NODE_ENV is not production ([#33](https://github.com/heroku/nodejs-npm-buildpack/pull/33))
- Warn when node modules are checked into git ([#34](https://github.com/heroku/nodejs-npm-buildpack/pull/34))
- Add logging method for warnings ([#35](https://github.com/heroku/nodejs-npm-buildpack/pull/35))
### Fixed
- Move integration testing to CirleCI ([#37](https://github.com/heroku/nodejs-npm-buildpack/pull/37))

## [0.2.0] 2020/05/19
### Added
- docs: add docs around `Permission denied` issues ([#28](https://github.com/heroku/nodejs-npm-buildpack/pull/28))
- Add dockerized unit tests ([#29](https://github.com/heroku/nodejs-npm-buildpack/pull/29))
- Added `provides` and `requires` of `node_modules` and `node` to buildplan. ([#18](https://github.com/heroku/nodejs-npm-buildpack/pull/18))

## [0.1.4] 2020/02/19
### Added
- feat: install `npm` version specified in `package.json` ([#24](https://github.com/heroku/nodejs-npm-buildpack/pull/24))
- feat: exchange echo commands for `log_info` method ([#25](https://github.com/heroku/nodejs-npm-buildpack/pull/25))
### Fixed
- fix: use_npm_ci expression return value id ([#22](https://github.com/heroku/nodejs-npm-buildpack/pull/23))

## [0.1.3] 2020/01/28
### Fixed
- fix: remove `-buildpack` from buildpack id ([#16](https://github.com/heroku/nodejs-npm-buildpack/pull/16))
- feat: support running on `io.buildpacks.stacks.bionic` stack ([#17](https://github.com/heroku/nodejs-npm-buildpack/pull/17))

## [0.1.2] 2019/11/01
### Added
- feat: support build time environment variables ([#14](https://github.com/heroku/nodejs-npm-buildpack/pull/14))

## [0.1.1] 2019/10/30
### Fixed
- Fix copying node_modules when a `package-lock.json` is present ([#12](https://github.com/heroku/nodejs-npm-buildpack/pull/12))

## [0.1.0] 2019/10/29
### Added
- feat: use `npm start` as the default launch.toml ([#11](https://github.com/heroku/nodejs-npm-buildpack/pull/11))

## [0.0.2] 2019/10/11
### Fixed
- Fix broken builds when a `package-lock.json` is missing ([#9](https://github.com/heroku/nodejs-npm-buildpack/pull/9))
