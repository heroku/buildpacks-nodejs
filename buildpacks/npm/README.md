# Node.js NPM Cloud Native Buildpack

This buildpack builds on top of the existing [Node.js Engine Cloud Native Buildpack](https://github.com/heroku/nodejs-engine-buildpack). It runs subsequent scripts after Node is install.

- Run automatically
  - `npm install` or `npm ci`
- Run when configured in `package.json`
  - `npm run build` or `npm run heroku-postbuild`

## Usage

### Install pack

Using `brew` (assuming development is done on MacOS), install `pack`.

```sh
brew tap buildpack/tap
brew install pack
```

If you're using Windows or Linux, follow instructions [here](https://buildpacks.io/docs/install-pack/).

### Install shpec (optional)

This buildpack uses `shpec` for unit tests, so to run them locally, you'll need to install the package.

```sh
curl -sLo- http://get.bpkg.sh | bash
bpkg install rylnd/shpec
```

### Clone the buildpack

Right now, we are prototyping with a local version of the buildpack. Clone it to your machine.

```sh
git clone git@github.com:heroku/nodejs-npm-buildpack.git
```

Clone the Heroku Node.js Engine Cloud Native Buildpack.

```sh
cd .. # change from nodejs-npm-buildpack directory
git clone git@github.com:heroku/nodejs-engine-buildpack.git
```

### Build the image

#### with buildpacks

Using pack, you're ready to create an image from the buildpack and source code. You will need to add flags that point to the path of the source code (`--path`) and the paths of the buildpacks (`--buildpack`).

```sh
cd nodejs-npm-buildpack
pack build TEST_IMAGE_NAME --path ../TEST_REPO_PATH --buildpack ../nodejs-engine-buildpack --buildpack ../nodejs-npm-buildpack
```

#### with a builder

You can also create a `builder.toml` file that will have explicit directions when creating a buildpack. This is useful when there are multiple "detect" paths a build can take (ie. yarn vs. npm commands).

In a directory outside of this buildpack, create a builder file:

```sh
cd ..
mkdir heroku_nodejs_builder
touch heroku_nodejs_builder/builder.toml
```

For local development, you'll want the file to look like this:

```toml
[[buildpacks]]
  id = "heroku/nodejs-engine"
  uri = "../nodejs-engine-buildpack"

[[buildpacks]]
  id = "heroku/nodejs-npm"
  uri = "../nodejs-npm-buildpack"

[[order]]
  group = [
    { id = "heroku/nodejs-engine", version = "0.4.3" },
    { id = "heroku/nodejs-npm", version = "0.1.4" }
  ]

[stack]
  id = "heroku-22"
  build-image = "heroku/heroku:22-cnb-build"
  run-image = "heroku/heroku:22-cnb"
```

Create the builder with `pack`:

```sh
pack create-builder nodejs --config ../heroku-nodejs-builder/builder.toml
```

Now you can use the builder image instead of chaining the buildpacks.

```sh
pack build TEST_IMAGE_NAME --path ../TEST_REPO_PATH --builder nodejs
```

### Common Issues

#### `jq: Permission denied` on a build

This issue may happen if a binary that is installed is not executable. This may happen on a Linux machine or while using a private network, such as a VPN, when using a local buildpack tool. If using `sfdx evergreen` or `pack`, pass in `--network host` to the command.

An example of this command running from the source code directory with a local builder image called `nodejs` would look like this:

```sh
pack build TEST_IMAGE_NAME --builder nodejs --network host
```

If building a function with `sfdx`, a command looks like this:

```sh
sfdx evergreen:functions:build image-repo/myfunction:dev --network host
```

## Testing

The complete test suite needs Docker to run. Make sure to [install Docker first](https://hub.docker.com/search?type=edition&offering=community).

```sh
make test
```

If you want to run individual test suites, that's available too.

**Unit Tests**

To run the tests on the local host, [make sure `shpec` is installed](#install-shpec-optional).

```sh
make unit-test
```

### Unit tests in Docker

Running the `shpec` aren't ideal since the test scripts read and write to the local buildpack directory, so Docker may be preferred.

As suggested above, install [Docker](#testing). Next, run the tests with the Make script:

```sh
make docker-unit-test
```

### Debugging tests

To debug, make changes from the code and rerun with the make command. To see what is happening, I suggest wrapping code blocks in question with `set -x`/`set +x`. It would look like this in the shpec file:

```sh
set -x
it "creates a toolbox.toml"
  install_or_reuse_toolbox "$layers_dir/toolbox"

  assert file_present "$layers_dir/toolbox.toml"
end
set +x
```

## Contributing

1. Open a pull request.
2. Make update to `CHANGELOG.md` under `main` with a description (PR title is fine) of the change, the PR number and link to PR.
3. Let the tests run on CI. When tests pass and PR is approved, the branch is ready to be merged.
4. Merge branch to `main`.

### Release

Note: if you're not a contributor to this project, a contributor will have to make the release for you.

1. Create a new branch (ie. `1.14.2-release`).
2. Update the version in the `buildpack.toml`.
3. Move the changes from `main` to a new header with the version and date (ie. `1.14.2 (2020-02-30)`).
4. Open a pull request.
5. Let the tests run on CI. When tests pass and PR is approved, the branch is ready to be merged.
6. Merge branch to `main`.
7. Pull down `main` to local machine.
8. Tag the current `main` with the version. (`git tag v1.14.2`)
9. Push up to GitHub. (`git push origin main --tags`) CI will run the suite and create a new release on successful run.

## Glossary

- buildpacks: provide framework and a runtime for source code. Read more [here](https://buildpacks.io).
- OCI image: [OCI (Open Container Initiative)](https://www.opencontainers.org/) is a project to create open sourced standards for OS-level virtualization, most importantly in Linux containers.
