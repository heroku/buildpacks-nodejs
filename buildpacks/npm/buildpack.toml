api = "0.2"

[buildpack]
id = "heroku/nodejs-npm"
version = "0.4.5"
name = "NPM Buildpack"
homepage = "https://github.com/heroku/buildpacks-nodejs"
keywords = ["nodejs", "node", "npm"]

[[buildpack.licenses]]
type = "MIT"

[[stacks]]
id = "heroku-18"

[[stacks]]
id = "heroku-20"

[[stacks]]
id = "io.buildpacks.stacks.bionic"

[metadata]

[metadata.release]

[metadata.release.docker]
repository = "public.ecr.aws/heroku-buildpacks/heroku-nodejs-npm-buildpack"
