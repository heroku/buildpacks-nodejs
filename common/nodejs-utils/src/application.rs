use crate::package_manager::PackageManager;
use bullet_stream::style;
use indoc::{formatdoc, writedoc};
use std::fmt::{Display, Formatter};
use std::path::Path;

/// Checks for npm, Yarn, pnpm, and shrink-wrap lockfiles and raises an error if multiple are detected.
///
/// # Errors
///
/// Will return an `Err` when:
/// - More than one lockfile exists in the `app_dir`.
/// - No lockfile exists in the `app_dir`.
pub fn check_for_singular_lockfile(app_dir: &Path) -> Result<(), Error> {
    let detected_lockfiles = [
        PackageManager::Npm,
        PackageManager::Pnpm,
        PackageManager::Yarn,
    ]
    .into_iter()
    .filter(|package_manager| app_dir.join(package_manager.lockfile()).exists())
    .collect::<Vec<_>>();

    match detected_lockfiles.len() {
        0 => Err(Error::MissingLockfile),
        1 => Ok(()),
        _ => Err(Error::MultipleLockfiles(detected_lockfiles)),
    }
}

/// Checks if the `node_modules` folder is present in the given directory which indicates that
/// the application contains files that it shouldn't in its git repository. If this is the case,
/// a warning message will be returned.
#[must_use]
pub fn warn_prebuilt_modules(app_dir: &Path) -> Option<String> {
    if app_dir.join("node_modules").exists() {
        Some(formatdoc! {"
            Warning: {node_modules} checked into source control

            Add these files and directories to {gitignore}. See the Dev Center for more info:
            https://devcenter.heroku.com/articles/node-best-practices#only-git-the-important-bits
        ", node_modules = style::value("node_modules"), gitignore = style::value(".gitignore") })
    } else {
        None
    }
}

#[derive(Debug)]
pub enum Error {
    MissingLockfile,
    MultipleLockfiles(Vec<PackageManager>),
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::MissingLockfile => {
                writedoc!(
                    f,
                    "
                        Couldn't determine Node.js package manager. Package \
                        manager lockfile not found.

                        A lockfile from a supported package manager is \
                        required to install Node.js dependencies. The \
                        package.json for this project specifies dependencies, \
                        but there isn't a lockfile.

                        To use npm to install dependencies, run {npm_install}. \
                        This command will generate a {npm_lockfile} lockfile.

                        Or, to use yarn to install dependencies, run {yarn_install}. \
                        This command will generate a {yarn_lockfile} lockfile.

                        Or, to use pnpm to install dependencies, run {pnpm_install}. \
                        This command will generate a {pnpm_lockfile} lockfile.

                        Ensure the resulting lockfile is committed to the repository, then try again.
                    ",
                    npm_install = style::value("npm install"),
                    npm_lockfile = style::value(PackageManager::Npm.lockfile().to_string_lossy()),
                    yarn_install = style::value("yarn install"),
                    yarn_lockfile = style::value(PackageManager::Yarn.lockfile().to_string_lossy()),
                    pnpm_install = style::value("pnpm install"),
                    pnpm_lockfile = style::value(PackageManager::Pnpm.lockfile().to_string_lossy()),
                )?;
                Ok(())
            }
            Error::MultipleLockfiles(package_managers) => {
                let lockfiles = package_managers
                    .iter()
                    .map(|f| f.lockfile().to_string_lossy().to_string())
                    .collect::<Vec<_>>()
                    .join(", ");

                writedoc!(f, "
                    Multiple lockfiles found: {lockfiles}

                    More than one package manager has created lockfiles for this application. Only one \
                    can be used to install dependencies but the buildpack can't determine which when multiple \
                    lockfiles are present.

                ")?;

                for package_manager in PackageManager::iterator() {
                    writedoc!(f, "- To use {} to install your application's dependencies please delete the following lockfiles:\n\n", style::value(package_manager.to_string()))?;
                    for other_package_manager in PackageManager::iterator() {
                        if package_manager != other_package_manager {
                            let other_lockfile = other_package_manager
                                .lockfile()
                                .to_string_lossy()
                                .to_string();
                            writedoc!(f, "    $ git rm {other_lockfile}\n")?;
                        }
                    }
                    writedoc!(f, "\n")?;
                }
                writedoc!(
                    f, "
                        See the Knowledge Base for more info: https://help.heroku.com/0KU2EM53

                        Once your application has only one lockfile, commit the results to git and retry your build.
                    "
                )?;

                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_output_for_multiple_lockfiles() {
        let error = Error::MultipleLockfiles(vec![
            PackageManager::Npm,
            PackageManager::Pnpm,
            PackageManager::Yarn,
        ]);
        assert_eq!(
            error.to_string(),
            formatdoc! {"
                Multiple lockfiles found: package-lock.json, pnpm-lock.yaml, yarn.lock

                More than one package manager has created lockfiles for this application. Only one can be used to install dependencies but the buildpack can't determine which when multiple lockfiles are present.

                - To use {npm} to install your application's dependencies please delete the following lockfiles:

                    $ git rm pnpm-lock.yaml
                    $ git rm yarn.lock

                - To use {pnpm} to install your application's dependencies please delete the following lockfiles:

                    $ git rm package-lock.json
                    $ git rm yarn.lock

                - To use {yarn} to install your application's dependencies please delete the following lockfiles:

                    $ git rm package-lock.json
                    $ git rm pnpm-lock.yaml

                See the Knowledge Base for more info: https://help.heroku.com/0KU2EM53

                Once your application has only one lockfile, commit the results to git and retry your build.
            ", npm = style::value("npm"), pnpm = style::value("pnpm"), yarn = style::value("Yarn") }
        );
    }
}
