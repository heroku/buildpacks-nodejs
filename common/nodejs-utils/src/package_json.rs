use crate::vrs::{Requirement, Version};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use thiserror::Error;

#[derive(Deserialize, Debug, Clone)]
pub struct PackageJson {
    pub name: String,
    pub version: Option<Version>,
    pub engines: Option<Engines>,
    pub scripts: Option<Scripts>,
    pub main: Option<String>,
    pub dependencies: Option<HashMap<String, String>>,
    #[serde(rename = "devDependencies")]
    pub dev_dependencies: Option<HashMap<String, String>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Engines {
    pub node: Option<Requirement>,
    pub yarn: Option<Requirement>,
    pub npm: Option<Requirement>,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct Scripts {
    pub start: Option<String>,
    pub build: Option<String>,
    pub heroku_prebuild: Option<String>,
    pub heroku_build: Option<String>,
    pub heroku_postbuild: Option<String>,
}

#[derive(Error, Debug)]
pub enum PackageJsonError {
    #[error("Could not read package.json. {0}")]
    AccessError(std::io::Error),
    #[error("Could not parse package.json. {0}")]
    ParseError(serde_json::Error),
}

impl PackageJson {
    /// Reads a package.json from a path into a `PackageJson` struct.
    ///
    /// # Errors
    ///
    /// * Invalid/malformed JSON
    /// * Path does not exist or is unreadable
    /// * Version strings are invalid/malformed
    pub fn read<P: AsRef<Path>>(path: P) -> Result<Self, PackageJsonError> {
        let file = File::open(path).map_err(PackageJsonError::AccessError)?;
        let rdr = BufReader::new(file);
        serde_json::from_reader(rdr).map_err(PackageJsonError::ParseError)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::Builder;

    #[test]
    fn read_valid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{\"name\": \"foo\",\"version\": \"0.0.0\"}}").unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(pkg.name, "foo");
        assert_eq!(pkg.version.unwrap().to_string(), "0.0.0");
    }

    #[test]
    fn read_valid_package_with_node_engine() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(
            f,
            "{{
            \"name\": \"foo\",
            \"version\": \"0.0.0\",
            \"engines\": {{
                \"node\": \"16.0.0\"
            }}
        }}"
        )
        .unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(&pkg.engines.unwrap().node.unwrap().to_string(), "16.0.0");
    }

    #[test]
    fn read_missing_package() {
        let res = PackageJson::read(Path::new("/over/there/package.json"));
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not read package.json"));
    }

    #[test]
    fn read_invalid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{").unwrap();
        let res = PackageJson::read(f.path());
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not parse package.json"));
    }
}
