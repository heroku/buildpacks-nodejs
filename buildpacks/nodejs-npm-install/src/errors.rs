use crate::npm;
use crate::BUILDPACK_NAME;
use commons::output::build_log::{BuildLog, Logger, StartedLogger};
use commons::output::fmt;
use commons::output::fmt::DEBUG_INFO;
use fun_run::CmdError;
use heroku_nodejs_utils::application;
use heroku_nodejs_utils::package_json::PackageJsonError;
use indoc::formatdoc;
use std::fmt::Display;
use std::io;
use std::io::stdout;

const USE_DEBUG_INFORMATION_AND_RETRY_BUILD: &str = "\
Use the debug information above to troubleshoot and retry your build.";

const SUBMIT_AN_ISSUE: &str = "\
If the issue persists and you think you found a bug in the buildpack then reproduce the issue \
locally with a minimal example and open an issue in the buildpack's GitHub repository with the details.";

#[derive(Debug)]
pub(crate) enum NpmInstallBuildpackError {
    Application(application::Error),
    BuildScript(CmdError),
    Detect(io::Error),
    NpmInstall(CmdError),
    NpmSetCacheDir(CmdError),
    NpmVersion(npm::VersionError),
    PackageJson(PackageJsonError),
}

pub(crate) fn on_error(error: libcnb::Error<NpmInstallBuildpackError>) {
    let logger = BuildLog::new(stdout()).without_buildpack_name();
    match error {
        libcnb::Error::BuildpackError(buildpack_error) => {
            on_buildpack_error(buildpack_error, logger);
        }
        framework_error => on_framework_error(&framework_error, logger),
    }
}

fn on_buildpack_error(error: NpmInstallBuildpackError, logger: Box<dyn StartedLogger>) {
    match error {
        NpmInstallBuildpackError::Application(e) => on_application_error(&e, logger),
        NpmInstallBuildpackError::BuildScript(e) => on_build_script_error(&e, logger),
        NpmInstallBuildpackError::Detect(e) => on_detect_error(&e, logger),
        NpmInstallBuildpackError::NpmInstall(e) => on_npm_install_error(&e, logger),
        NpmInstallBuildpackError::NpmSetCacheDir(e) => on_set_cache_dir_error(&e, logger),
        NpmInstallBuildpackError::NpmVersion(e) => on_npm_version_error(e, logger),
        NpmInstallBuildpackError::PackageJson(e) => on_package_json_error(e, logger),
    }
}

fn on_package_json_error(error: PackageJsonError, logger: Box<dyn StartedLogger>) {
    match error {
        PackageJsonError::AccessError(e) => {
            print_error_details(logger, &e)
                .announce()
                .error(&formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file can’t be read. 
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = fmt::value("package.json") });
        }
        PackageJsonError::ParseError(e) => {
            print_error_details(logger, &e)
                .announce()
                .error(&formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file \
                    can’t be parsed. Ensure {npm_install} runs locally to check the formatting in your file.
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = fmt::value("package.json"), npm_install = fmt::value("npm install") });
        }
    }
}

fn on_set_cache_dir_error(error: &CmdError, logger: Box<dyn StartedLogger>) {
    print_error_details(logger, &error)
        .announce()
        .error(&formatdoc! {"
            Failed to set the {npm} cache directory.

            An unexpected error occurred while setting the {npm} cache directory. 
                    
            {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

            {SUBMIT_AN_ISSUE}
        ", npm = fmt::value("npm") });
}

fn on_npm_version_error(error: npm::VersionError, logger: Box<dyn StartedLogger>) {
    match error {
        npm::VersionError::Command(e) => {
            print_error_details(logger, &e)
                .announce()
                .error(&formatdoc! {"
                    Failed to determine {npm} version information.

                    An unexpected error occurred while executing {npm_version}.  
                    
                    {SUBMIT_AN_ISSUE}
                ", npm = fmt::value("npm"), npm_version = fmt::value(e.name()) });
        }
        npm::VersionError::Parse(stdout, e) => {
            print_error_details(logger, &e)
                .announce()
                .error(&formatdoc! {"
                    Failed to parse {npm} version information.
        
                    An unexpected error occurred while parsing version information from {output}. 
                    
                    {SUBMIT_AN_ISSUE}
                ", npm = fmt::value("npm"), output = fmt::value(stdout) });
        }
    }
}

fn on_npm_install_error(error: &CmdError, logger: Box<dyn StartedLogger>) {
    print_error_details(logger, &error)
        .announce()
        .error(&formatdoc! {"
            Failed to install Node modules.

            The {buildpack_name} uses the command {npm_install} to install your Node modules. This command \
            failed and the buildpack cannot continue. See the log output above for more information.

            This error can occur due to an unstable network connection. Ensure that this command runs locally \
            without error (exit status = 0) and retry your build.
            
            If that doesn’t help, check the status of the upstream Node module repository service at https://status.npmjs.org/.
        ", npm_install = fmt::value(error.name()), buildpack_name = fmt::value(BUILDPACK_NAME) });
}

fn on_build_script_error(error: &CmdError, logger: Box<dyn StartedLogger>) {
    print_error_details(logger, &error)
        .announce()
        .error(&formatdoc! {"
            Failed to execute build script.

            The {buildpack_name} allows customization of the build process by executing the following scripts \
            if they are defined in {package_json}:
            - {heroku_prebuild} 
            - {heroku_build} or {build} 
            - {heroku_postbuild}

            An unexpected error occurred while executing {build_script}. See the log output above for more information.

            Ensure that this command runs locally without error and retry your build.
        ",
            build_script = fmt::value(error.name()),
            buildpack_name = fmt::value(BUILDPACK_NAME),
            package_json = fmt::value("package.json"),
            heroku_prebuild = fmt::value("heroku-prebuild"),
            heroku_build = fmt::value("heroku-build"),
            build = fmt::value("build"),
            heroku_postbuild = fmt::value("heroku-postbuild"),
        });
}

fn on_application_error(error: &application::Error, logger: Box<dyn StartedLogger>) {
    logger.announce().error(&error.to_string());
}

fn on_detect_error(error: &io::Error, logger: Box<dyn StartedLogger>) {
    print_error_details(logger, &error)
        .announce()
        .error(&formatdoc! {"
            Unable to complete buildpack detection.

            An unexpected error occurred while determining if the {buildpack_name} should be \
            run for this application. See the log output above for more information. 
        ", buildpack_name = fmt::value(BUILDPACK_NAME) });
}

fn on_framework_error(
    error: &libcnb::Error<NpmInstallBuildpackError>,
    logger: Box<dyn StartedLogger>,
) {
    print_error_details(logger, &error)
        .announce()
        .error(&formatdoc! {"
            {buildpack_name} internal error.

            The framework used by this buildpack encountered an unexpected error.
            
            If you can't deploy to Heroku due to this issue, check the official Heroku Status page at \
            status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

            {SUBMIT_AN_ISSUE}
        ", buildpack_name = fmt::value(BUILDPACK_NAME) });
}

fn print_error_details(
    logger: Box<dyn StartedLogger>,
    error: &impl Display,
) -> Box<dyn StartedLogger> {
    logger
        .section(DEBUG_INFO)
        .step(&error.to_string())
        .end_section()
}
