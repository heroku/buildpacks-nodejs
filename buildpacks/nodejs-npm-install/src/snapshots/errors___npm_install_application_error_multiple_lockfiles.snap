---
source: buildpacks/nodejs-npm-install/src/errors.rs
---
! Multiple lockfiles detected
!
! The following lockfiles were found in this application: package-lock.json, yarn.lock, pnpm-lock.yaml
!
! More than one package manager has created lockfiles for this application. Only one can be used to install dependencies but the buildpack can't determine which when multiple lockfiles are present.
!
! - To use `npm` to install your application's dependencies please delete the following lockfiles:
!
!     $ git rm pnpm-lock.yaml
!     $ git rm yarn.lock
!
! - To use `pnpm` to install your application's dependencies please delete the following lockfiles:
!
!     $ git rm package-lock.json
!     $ git rm yarn.lock
!
! - To use `Yarn` to install your application's dependencies please delete the following lockfiles:
!
!     $ git rm package-lock.json
!     $ git rm pnpm-lock.yaml
!
! See the Knowledge Base for more info: https://help.heroku.com/0KU2EM53
!
! Once your application has only one lockfile, commit the results to git and retry your build.
